### PR TITLE
Implement JSON object

### PR DESCRIPTION
**Pull request by rtakacs**
JerryScript-DCO-1.0-Signed-off-by: Zoltan Herczeg zherczeg.u-szeged@partner.samsung.com
JerryScript-DCO-1.0-Signed-off-by: Roland Takacs rtakacs.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 1]
```diff
@@ -0,0 +1,1840 @@
...
+  /* 6. */
+  else
+  {
+    for (ecma_property_t *property_p = ecma_get_property_list (obj_p);
+         property_p != NULL;
+         property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+    {
+      if (property_p->type != ECMA_PROPERTY_NAMEDDATA || !ecma_is_property_enumerable (property_p))
+      {
+        continue;
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Done.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 2]
```diff
@@ -0,0 +1,1840 @@
...
+      value = ecma_copy_value (val, true);
+
+      ECMA_FINALIZE (val);
+    }
+  }
+
+  if (!ecma_is_completion_value_empty (ret_value))
+  {
+    ecma_free_value (value, true);
+    return ret_value;
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Yes we could. I updated this function to have only one return at the end.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 3]
```diff
@@ -0,0 +1,1840 @@
...
+                       ecma_object_t *holder_p, /**< the object*/
+                       stringify_context_t *context_p) /**< context*/
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+  ecma_value_t value;
+
+  /* 1. */
+  ecma_completion_value_t get_comp_val = ecma_op_object_get (holder_p, key_p);
+
+  if (!ecma_is_completion_value_normal (get_comp_val))
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Fixed.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 4]
```diff
@@ -0,0 +1,1840 @@
...
+  {
+    return get_comp_val;
+  }
+
+  value = ecma_copy_value (ecma_get_completion_value_value (get_comp_val), true);
+  ecma_free_completion_value (get_comp_val);
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: In this case I could avoid some (early) returns from an ECMA_TRY_CATCH. But now I use only one return at the end.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+  size_t old_block_size = static_cast<size_t> (ctx_p->block_end_p - ctx_p->block_start_p);
+  size_t new_block_size = old_block_size + LIST_BLOCK_SIZE;
+
+  size_t current_ptr_offset = static_cast<size_t> (ctx_p->current_p - ctx_p->block_start_p);
+  void **new_block_start_p = (void **) mem_heap_alloc_block (new_block_size, MEM_HEAP_ALLOC_SHORT_TERM);
+
+  if (ctx_p->current_p)
+  {
+    memcpy (new_block_start_p, ctx_p->block_start_p, static_cast<size_t> (current_ptr_offset));
+    mem_heap_free_block (ctx_p->block_start_p);
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp]: Yes, maybe it's better to use the ecma_collection with an append functionality.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+
+#define LIST_BLOCK_SIZE 256UL
+
+/**
+ * Allocate memory for elements.
+ *
+ * @return pointer to current position of the allocated memory.
+ */
+void **
+realloc_list (list_ctx_t *ctx_p) /**< list context */
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp]: Yes.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+ *
+ * \addtogroup ecmahelpers Helpers for operations with ECMA data types
+ * @{
+ */
+
+#include "ecma-alloc.h"
+#include "ecma-helpers.h"
+#include "ecma-builtin-helpers.h"
+
+#define LIST_BLOCK_SIZE 256UL
```
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers-json.cpp]: This is a copy from the regexp engine. Maybe a bit big. :)

**Comment by zherczeg** [comment]
Since the project went on public, this is not updated automatically anymore.

**Comment by zherczeg** [comment]
New PR created: #307

**Pull request closed without merging**